File: solutions/2187. Minimum Time to Complete Trips/2187.cpp
class Solution {
 public:
  long long minimumTime(vector<int>& time, int totalTrips) {
    long long l = 1;
    long long r =
        *min_element(begin(time), end(time)) * static_cast<long>(totalTrips);

    auto canComplete = [&](int m) -> bool {
      long long trips = 0;
      for (const int t : time)
        trips += m / t;
      return trips >= totalTrips;
    };

    while (l < r) {
      const long long m = (l + r) / 2;
      if (canComplete(m))
        r = m;
      else
        l = m + 1;
    }

    return l;
  }
};

File: solutions/2187. Minimum Time to Complete Trips/2187.java
class Solution {
  public long minimumTime(int[] time, int totalTrips) {
    long l = 1;
    long r = Arrays.stream(time).min().getAsInt() * (long) totalTrips;

    while (l < r) {
      final long m = (l + r) / 2;
      if (canComplete(time, totalTrips, m))
        r = m;
      else
        l = m + 1;
    }

    return l;
  }

  private boolean canComplete(int[] time, int totalTrips, long m) {
    long trips = 0;
    for (final int t : time)
      trips += m / t;
    return trips >= totalTrips;
  }
}
